src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs

```diff
@@ -0,0 +1,1284 @@
+
+            comp = CreateCompilation(src3);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TypeDeclaration_IsStruct()
+        {
+            var src = @"
+record struct Point(int x, int y);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+
+            var point = comp.GlobalNamespace.GetTypeMember("Point");
+            Assert.True(point.IsValueType);
+            Assert.False(point.IsReferenceType);
+            Assert.False(point.IsRecord);
+            Assert.True(point.IsRecordStruct);
+            Assert.Equal(TypeKind.Struct, point.TypeKind);
+            Assert.Equal(SpecialType.System_ValueType, point.BaseTypeNoUseSiteDiagnostics.SpecialType);
+
+            Assert.True(SyntaxFacts.IsTypeDeclaration(SyntaxKind.RecordStructDeclaration));
+        }
+
+        [Fact]
+        public void TypeDeclaration_MayNotHaveBaseType()
+        {
+            var src = @"
+record struct Point(int x, int y) : object;
+record struct Point2(int x, int y) : System.ValueType;
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,37): error CS0527: Type 'object' in interface list is not an interface
+                // record struct Point(int x, int y) : object;
+                Diagnostic(ErrorCode.ERR_NonInterfaceInInterfaceList, "object").WithArguments("object").WithLocation(2, 37),
+                // (3,38): error CS0527: Type 'ValueType' in interface list is not an interface
+                // record struct Point2(int x, int y) : System.ValueType;
+                Diagnostic(ErrorCode.ERR_NonInterfaceInInterfaceList, "System.ValueType").WithArguments("System.ValueType").WithLocation(3, 38)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_MayNotHaveTypeConstraintsWithoutTypeParameters()
+        {
+            var src = @"
+record struct Point(int x, int y) where T : struct;
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (2,35): error CS0080: Constraints are not allowed on non-generic declarations
+                // record struct Point(int x, int y) where T : struct;
+                Diagnostic(ErrorCode.ERR_ConstraintOnlyAllowedOnGenericDecl, "where").WithLocation(2, 35)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_AllowedModifiers()
+        {
+            var src = @"
+readonly partial record struct S1;
+public record struct S2;
+internal record struct S3;
+
+public class Base
+{
+    public int S6;
+}
+public class C : Base
+{
+    private protected record struct S4;
+    protected internal record struct S5;
+    new record struct S6;
+}
+unsafe record struct S7;
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (16,22): error CS0227: Unsafe code may only appear if compiling with /unsafe
```

Comment:
>// (16,22): error CS0227: Unsafe code may only appear if compiling with /unsafe [](start = 16, length = 79)

It looks like this error is covered in the test below. I was looking for a success case.

Reply:
Ah. Changed to use "unsafe" compilation option

---
In reply to: [588363595](https://github.com/dotnet/roslyn/pull/51494#discussion_r588363595) [](ancestors = 588363595)